File: Java/886.java
__________________________________________________________________________________________________
sample 3 ms submission
class Solution {
    public boolean possibleBipartition(int N, int[][] dislikes) {
        int[] parent = new int[N+1];
        for (int[] d : dislikes){
            if (parent[d[1]] == 0)
                parent[d[1]] = d[0]; 
            else {
                boolean c0 = false;
                int a = d[0];               //a = hater
                while(parent[a] != 0){      //if somebody already hates a
                    a = parent[a];          //set a to be the one who hates a
                    c0 = c0 ? false : true; //odd or even length of hate chain
                }
                boolean c1 = false;
                int b = d[1];               //same as above
                while(parent[b] != 0){
                    b = parent[b];
                    c1 = c1 ? false : true;
                }
                if (a == b && c0 == c1) return false; //same beginning of hate chain && same odd/even chain length leads to false
            }
        }
        return true;
    }
}
__________________________________________________________________________________________________
sample 58084 kb submission
class Solution {
    private ArrayList<Integer>[] graph;
    private Map<Integer, Integer> color;
    public boolean possibleBipartition(int N, int[][] dislikes) {
        graph = new ArrayList[N+1];
        for (int i =0; i<=N;i++) {
            graph[i] = new ArrayList<>();
        }
        for(int[] dis: dislikes) {
            graph[dis[0]].add(dis[1]);
            graph[dis[1]].add(dis[0]);
        }
        color = new HashMap<>();
        for(int i =1;i<=N;i++) {
            if(!color.containsKey(i) && !dfs(i, 0)) {
                return false;
            }
        }
        return true;
    }
    private boolean dfs(int node, int code) {
        if (color.containsKey(node)) {
            return color.get(node)%2 == code%2;
        }
        color.put(node, code);
        for(int nei: graph[node]) {
            if (!dfs(nei, code + 1)) return false;
        }
        return true;
    }
}
__________________________________________________________________________________________________

File: Python3/886.py
__________________________________________________________________________________________________
sample 772 ms submission
class Solution:
    def possibleBipartition(self, N: int, dislikes: List[List[int]]) -> bool:
        d = collections.defaultdict(list)
        for a, b in dislikes:
            d[a].append(b)
            d[b].append(a)
        color = [0] * (N + 1)
        for i in range(1, N + 1):
            if not color[i]:
                queue = collections.deque([i])
                color[i] = 1
                while queue:
                    cur = queue.popleft()
                    for nei in d[cur]:
                        if not color[nei]:
                            color[nei] = -color[cur]
                            queue.append(nei)
                        elif color[nei] == color[cur]:
                            return False
        return True
__________________________________________________________________________________________________
sample 784 ms submission
from collections import defaultdict


class Solution:
    def possibleBipartition(self, n: int, dislikes: List[List[int]]) -> bool:
        graph = defaultdict(list)
        for a, b in dislikes:
            graph[a].append(b)
            graph[b].append(a)
        
        tags = {}
        for start in range(1, n+1):
            if start in tags:
                continue
                
            tags[start] = 0
            to_visit = [start]
            while to_visit:
                node = to_visit.pop()
                tag = tags[node]
                for neighbor in graph[node]:
                    if neighbor not in tags:
                        tags[neighbor] = 1 - tag    # Alternates between 0 and 1
                        to_visit.append(neighbor)
                    else:
                        if tags[neighbor] != 1 - tag:
                            return False
        return True
__________________________________________________________________________________________________
